mllib/src/main/scala/org/apache/spark/ml/param/params.scala

```diff
@@ -311,7 +324,11 @@ trait Params extends Identifiable with Serializable {
    */
   protected final def set[T](param: Param[T], value: T): this.type = {
     shouldOwn(param)
-    paramMap.put(param.asInstanceOf[Param[Any]], value)
+    if (param.isInstanceOf[StringArrayParam] && value.isInstanceOf[Seq[_]]) {
```

Comment:
This is specialized for `StringArrayParam`. If user adds other `Param` types, they cannot modify the code. We can add `set` (or `setAll`) with varargs that takes `ParamPair[_]*`, and let users to create `ParamPair` first if the value type is different from the param type. For this function, `paramMap.put(param.w(value))` should be sufficient.